src/source_range.rs

```diff
@@ -0,0 +1,113 @@
+use super::{cmark_resume_one_event, fmt, Borrow, Event, Options, Range, State};
+
+/// Serialize a stream of [pulldown-cmark-Events][Event] while preserving the escape characters in `source`.
+/// Each input [Event] is accompanied by an optional [Range] that maps it back to the `source` string.
+///
+/// Different from [`cmark_resume_with_options`], which always escape Markdown special characters like `#` or `[`,
+/// this function only escapes a special character if it is escaped in `source`.
```

Comment:
Is there any way we could avoid adding escapes in the other `cmark` variants as well, or does this fundamentally require the source?

Reply:
@max-heller, this variant "preserves escapes", i.e., adds escapes based on whether the source adds escapes, so it requires the source.

Reply:
Should've been more explicit, sorry. Obviously we can't **preserve** escapes without the source, but could we prevent **adding them** without it? For instance, keep `[foo]` as `[foo]` but turn `\[foo\]` to `[foo]`? Or would that run in to correctness issues